File: gluecodium/src/test/resources/smoke/nesting/output/android/com/example/smoke/OuterStruct.java
/*
 *
 */
package com.example.smoke;
import android.support.annotation.NonNull;
import com.example.NativeBase;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
public final class OuterStruct {
    @NonNull
    public String field;
    public enum InnerEnum {
        FOO(0),
        BAR(1);
        public final int value;
        InnerEnum(final int value) {
            this.value = value;
        }
    }
    public static final class InstantiationException extends Exception {
        public InstantiationException(final OuterStruct.InnerEnum error) {
            super(error.toString());
            this.error = error;
        }
        public final OuterStruct.InnerEnum error;
    }
    public static final class InnerStruct {
        @NonNull
        public List<Date> otherField;
        public InnerStruct(@NonNull final List<Date> otherField) {
            this.otherField = otherField;
        }
        public native void doSomething();
    }
    public static final class InnerClass extends NativeBase {
        /**
         * For internal use only.
         * @hidden
         */
        protected InnerClass(final long nativeHandle, final Object dummy) {
            super(nativeHandle, new Disposer() {
                @Override
                public void disposeNative(long handle) {
                    disposeNativeHandle(handle);
                }
            });
        }
        private static native void disposeNativeHandle(long nativeHandle);
        @NonNull
        public native Set<Locale> fooBar();
    }
    public static final class Builder extends NativeBase {
        public Builder() {
            this(create(), (Object)null);
            cacheThisInstance();
        }
        /**
         * For internal use only.
         * @hidden
         */
        protected Builder(final long nativeHandle, final Object dummy) {
            super(nativeHandle, new Disposer() {
                @Override
                public void disposeNative(long handle) {
                    disposeNativeHandle(handle);
                }
            });
        }
        private static native void disposeNativeHandle(long nativeHandle);
        private native void cacheThisInstance();
        private static native long create();
        @NonNull
        public native OuterStruct.Builder field(@NonNull final String value);
        @NonNull
        public native OuterStruct build();
    }
    /**
     * @hidden
     */
    static class InnerInterfaceImpl extends NativeBase implements InnerInterface {
        protected InnerInterfaceImpl(final long nativeHandle, final Object dummy) {
            super(nativeHandle, new Disposer() {
                @Override
                public void disposeNative(long handle) {
                    disposeNativeHandle(handle);
                }
            });
        }
        private static native void disposeNativeHandle(long nativeHandle);
        @NonNull
        public native Map<String, byte[]> barBaz();
    }
    public interface InnerInterface {
        @NonNull
        Map<String, byte[]> barBaz();
    }
    public OuterStruct(@NonNull final String field) {
        this.field = field;
    }
    public native void doNothing() throws OuterStruct.InstantiationException;
}

File: gluecodium/src/test/resources/smoke/nesting/output/android-kotlin/com/example/smoke/OuterStruct.kt
/*

 *
 */

@file:JvmName("OuterStructExtensions")


package com.example.smoke

import com.example.NativeBase
import java.util.Date
import java.util.Locale

class OuterStruct {
    @JvmField var field: String

    enum class InnerEnum(@JvmField val value: Int) {
        FOO(0),
        BAR(1);
    }
    class InstantiationException(@JvmField val error: OuterStruct.InnerEnum) : Exception(error.toString())


    class InnerStruct {
        @JvmField var otherField: List<Date>



        constructor(otherField: List<Date>) {
            this.otherField = otherField
        }





        external fun doSomething() : Unit


    }

    class InnerClass : NativeBase {



        /**
         * For internal use only.
         * @suppress
         * @param nativeHandle The handle to resources on C++ side.
         * @param tag Tag used by callers to avoid overload resolution problems.
         */
        protected constructor(nativeHandle: Long, @Suppress("UNUSED_PARAMETER") tag: Any?)
            : super(nativeHandle, { disposeNativeHandle(it) }) {}





        external fun fooBar() : Set<Locale>




        companion object {
            @JvmStatic private external fun disposeNativeHandle(nativeHandle: Long)
        }
    }


    class Builder : NativeBase {




        constructor() : this(create(), null as Any?) {
            cacheThisInstance();
        }

        /**
         * For internal use only.
         * @suppress
         * @param nativeHandle The handle to resources on C++ side.
         * @param tag Tag used by callers to avoid overload resolution problems.
         */
        protected constructor(nativeHandle: Long, @Suppress("UNUSED_PARAMETER") tag: Any?)
            : super(nativeHandle, { disposeNativeHandle(it) }) {}

        private external fun cacheThisInstance()




        external fun field(value: String) : OuterStruct.Builder


        external fun build() : OuterStruct




        companion object {
            @JvmStatic private external fun disposeNativeHandle(nativeHandle: Long)

            @JvmStatic private external fun create() : Long
        }
    }


    fun interface InnerInterface {



        fun barBaz() : Map<String, ByteArray>


    }

    /**
     * @suppress
     *
     * This class is used to represent C++ implementations of the interface or lambda in Kotlin.
     * It is instantiated by JNI and should not be used by the end users.
     */
    private class InnerInterfaceImpl : NativeBase, InnerInterface {
        protected constructor(nativeHandle: Long, @Suppress("UNUSED_PARAMETER") tag: Any?)
            : super(nativeHandle, { disposeNativeHandle(it) }) {}



        override external fun barBaz() : Map<String, ByteArray>



        companion object {
            @JvmStatic private external fun disposeNativeHandle(nativeHandle: Long)
        }
    }
    fun interface InnerLambda {


        fun apply() : Unit
    }

    /**
     * @suppress
     *
     * This class is used to represent C++ implementations of the interface or lambda in Kotlin.
     * It is instantiated by JNI and should not be used by the end users.
     */
    private class InnerLambdaImpl : NativeBase, InnerLambda {
        protected constructor(nativeHandle: Long, @Suppress("UNUSED_PARAMETER") tag: Any?)
            : super(nativeHandle, { disposeNativeHandle(it) }) {}



        override external fun apply() : Unit



        companion object {
            @JvmStatic private external fun disposeNativeHandle(nativeHandle: Long)
        }
    }


    constructor(field: String) {
        this.field = field
    }




    @Throws(OuterStruct.InstantiationException::class)
    external fun doNothing() : Unit


}

